### ISSUE TITLE
Access to `model_type` from validators

### ISSUE THREAD

**Issue by dennispg** [pydantic/pydantic#11104]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

Currently, `field_name`, `data`, and `model_config` are provided in `ValidationInfo` of field validators, and where applicable have their values.

It would be really useful to also have `model_type` made available here as well.

```py
def validate_length(value: Any, info: ValidationInfo):
    #
    #   I want to see the json_schema_extra value from here
    #
    if info.field_name is not None and info.model_type is not None:
      field = info.model_type.model_fields[info.field_name]
        if field is not None and isinstance(field.json_schema_extra, dict):
          if field.json_schema_extra["absolute"]:
            return abs(value)
    return value

class SimpleThirdPartyType(BaseModel):
    length: int

AnnotatedModel = Annotated[SimpleThirdPartyType, BeforeValidator(validate_length)]

class SimpleOuterModel(BaseModel):
    value: AnnotatedModel = Field(json_schema_extra={"absolute": True})
```

Another example using `__get_pydantic_core_schema__`

```py
class _ThirdPartyTypePydanticAnnotation:
    @classmethod
    def __get_pydantic_core_schema__(
        cls,
        _source_type: Any,
        _handler: GetCoreSchemaHandler,
    ) -> core_schema.CoreSchema:
        def validate_from_int(value: int, info) -> ThirdPartyType:
            field = info.model_type.model_fields[info.field_name]
            #
            #   I want to read some metadata from the Field() call
            #
            result = ThirdPartyType()
            result.x = value
            return result

        from_int_schema = core_schema.chain_schema(
            [
                core_schema.int_schema(),
                core_schema.with_info_plain_validator_function(validate_from_int),
            ]
        )

        return core_schema.json_or_python_schema(
            json_schema=from_int_schema,
            python_schema=core_schema.union_schema(
                [
                    # check if it's an instance first before doing any further work
                    core_schema.is_instance_schema(ThirdPartyType),
                    from_int_schema,
                ]
            ),
            serialization=core_schema.plain_serializer_function_ser_schema(
                lambda instance: instance.x
            ),
        )

    @classmethod
    def __get_pydantic_json_schema__(
        cls, _core_schema: core_schema.CoreSchema, handler: GetJsonSchemaHandler
    ) -> JsonSchemaValue:
        # Use the same schema that would be used for `int`
        return handler(core_schema.int_schema())


# We now create an `Annotated` wrapper that we'll use as the annotation for fields on `BaseModel`s, etc.
PydanticThirdPartyType = Annotated[
    ThirdPartyType, _ThirdPartyTypePydanticAnnotation
]


# Create a model class that uses this annotation as a field
class Model(BaseModel):
    third_party_type: PydanticThirdPartyType = Field(title="third_party", json_schema_extra={"extra": "data"})
```

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [X] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [ ] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [X] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by dennispg** [comment]
Just to include a bit more of  an easier to read example:

```python
def validate_length(value: Any, info: ValidationInfo):
    #
    #   I want to see the json_schema_extra value from here
    #
    if info.field_name:
      field = info.model_fields[info.field_name)
        if field is not None and isinstance(field.json_schema_extra, dict):
          if field.json_schema_extra["absolute"]:
            return abs(value)
    return value

class SimpleThirdPartyType(BaseModel):
    length: int

AnnotatedModel = Annotated[SimpleThirdPartyType, BeforeValidator(validate_length)]

class SimpleOuterModel(BaseModel):
    value: AnnotatedModel = Field(json_schema_extra={"absolute": True})
```


On the other hand, the following works because the validator is a class method on the model itself, and so therefore has access to `cls` and therefore `cls.model_fields`. I don't have the same luxury when using annotations.

```python
class SimpleModel(BaseModel):
    length: int = Field(json_schema_extra={"absolute": True})

    @field_validator("length")
    def validate_length(cls, value: Any, info: ValidationInfo):
        if info.field_name:
            field = cls.model_fields.get(info.field_name)
            if field is not None and isinstance(field.json_schema_extra, dict):
                if field.json_schema_extra["absolute"]:
                    return abs(value)
        return value
```

**Comment by AdunSG** [comment]
fieldvalidator can be used as class method and therefore you can directly access model_fields from the class, or am i missing something?
```
    @field_validator('x', 'y')
    @classmethod
    def double(cls, v: str) -> str:
        for name, info in cls.model_fields:
            ...
```

**Comment by dennispg** [comment]
Even just the applicable `FieldInfo` from `model_fields` would be a big help if for any reason the full `model_fields` is a bridge to far to be able to provide.

**Comment by AdunSG** [comment]
Unfortunetly i'm not able to hep implement new features, but maybe you can elaborate on why creating PydanticThirdPartyType as BaseModel itself is no solution?
Nested BaseModels are well suported.

**Comment by dennispg** [comment]
Did a bit of digging to see how `field_name` was being provided, and found that there is a stack being kept for both `field_name` and `model_type`. I was able to use `handler._generate_schema.model_type_stack.get()` and get what I need.


```python
class SimpleModelAnnotation:
    @classmethod
    def __get_pydantic_core_schema__(
        cls,
        source_type: Any,
        handler: GetCoreSchemaHandler,
    ) -> core_schema.CoreSchema:
        #
        #   this works!
        #
        model_type = handler._generate_schema.model_type_stack.get() # type: ignore

        def validate_length(value: Any, info: ValidationInfo):
            if model_type is not None:
                field = model_type.model_fields[info.field_name] # type: ignore
                if field is not None and isinstance(field.json_schema_extra, dict):
                    if field.json_schema_extra["absolute"]:
                        value.length = abs(value.length)
                        return value
            return value

        third_party_schema = core_schema.with_info_before_validator_function(
            function=validate_length,
            schema=core_schema.is_instance_schema(SimpleThirdPartyType),
            field_name=handler.field_name,
        )

        return core_schema.json_or_python_schema(
            json_schema=third_party_schema,
            python_schema=third_party_schema
        )

class SimpleThirdPartyType(BaseModel):
    length: int

AnnotatedModel = Annotated[SimpleThirdPartyType, SimpleModelAnnotation]

class SimpleOuterModel(BaseModel):
    third_party_field: AnnotatedModel = Field(json_schema_extra={"absolute": True})
```


It looks trivially easy to add a model_type property to `CallbackGetCoreSchemaHandler` in `_internal/_schema_generation_shared.py`, but I couldn't yet figure out to get it into the `ValidationInfo` object.

I don't know if any of this goes against the intentions of how this should be used, but I'd be happy to submit a PR if it helps?

**Comment by Viicos** [comment]
It feels a bit weird to have the JSON Schema extras influencing the validation behavior. Instead, the JSON Schema should be generated according to the model definition. Here is one way to do it:

```python
from typing import Annotated

from pydantic import Field, ValidatorFunctionWrapHandler, WrapValidator


def validate_int(absolute: bool) -> WrapValidator:
    def validate(value: Any, handler: ValidatorFunctionWrapHandler) -> int:
        validated_value: int = handler(value)
        return abs(validated_value) if absolute else validated_value

    if absolute:
        return WrapValidator(
            validate, json_schema_input_type=Annotated[int, Field(json_schema_extra={'absolute': True})]
        )
    else:
        return WrapValidator(validate)


AbsoluteValidator = Annotated[int, validate_int(absolute=True)]

ta = TypeAdapter(AbsoluteValidator)
ta.validate_python(-1)
#> 1
ta.json_schema()
#> {'type': 'integer', 'absolute': True}
```

If using an annotation for a third-party type:

```python
from typing import Annotated

from pydantic import GetCoreSchemaHandler, GetJsonSchemaHandler, TypeAdapter
from pydantic.json_schema import JsonSchemaValue

from pydantic_core import core_schema


class ThirdPartyType:
    x: int

    def __repr__(self) -> str:
        return f'{self.__class__.__name__}(x={self.x})'


class ValidateThirdParty:
    def __init__(self, absolute: bool) -> None:
        self.absolute = absolute

    def __get_pydantic_core_schema__(
        self,
        _source_type: Any,
        _handler: GetCoreSchemaHandler,
    ) -> core_schema.CoreSchema:

        def validate_from_int(value: int) -> ThirdPartyType:
            result = ThirdPartyType()
            result.x = abs(value) if self.absolute else value
            return result

        from_int_schema = core_schema.chain_schema(
            [
                core_schema.int_schema(),
                core_schema.no_info_plain_validator_function(validate_from_int),
            ]
        )

        return core_schema.json_or_python_schema(
            json_schema=from_int_schema,
            python_schema=core_schema.union_schema(
                [
                    # check if it's an instance first before doing any further work
                    core_schema.is_instance_schema(ThirdPartyType),
                    from_int_schema,
                ]
            ),
            serialization=core_schema.plain_serializer_function_ser_schema(
                lambda instance: instance.x
            ),
        )

    def __get_pydantic_json_schema__(
        self,
        cs: core_schema.CoreSchema,
        handler: GetJsonSchemaHandler
    ) -> JsonSchemaValue:
        # Note: you can generate a different JSON Schema in serialization:
        # if handler.mode == 'serialization' and self.absolute:
        #     json_schema = handler(core_schema.int_schema(ge=0))
        json_schema = handler(core_schema.int_schema())
        if self.absolute:
            json_schema['absolute'] = True
        return json_schema


ta = TypeAdapter(Annotated[ThirdPartyType, ValidateThirdParty(absolute=True)])
```

**Comment by dennispg** [comment]
So I dervied that example as just a very simple one only to illustrate the ask, and I dont necessarily disagree on the awkwardness of using JSON Schema Extras. The real need though is that I need a way to be able to fill in arbitrary configuration on the model level, and using JSON Schema extras for this has the added benefit that it outputs to the JSON schema for me as well.

Here is one more specific example: I use the Pint library to manage quantities with units, and the `Quantity` object is a third party type that needs to be annotated to be used with pydantic models. Take this class for example:

```py
class Box(BaseModel):
    w: AnnotatedQuantity = Field(json_schema_extra={'units': 'in'})
    l: AnnotatedQuantity = Field(json_schema_extra={'units': 'in'})
    h: AnnotatedQuantity = Field(json_schema_extra={'units': 'in'})
```

I have classes like this all over that use all kinds of different units. The configured units influences how validation is done. Making a new annotation for every premutation of units feels unnecessary.

Again, this is just one example. Having access to the `model_type` feels like a fairly easy thing to expose and opens up a wealth of possibilities to users.

I have a PR ready to submit towards this if you agree about the value in it.

**Comment by Viicos** [comment]
> Making a new annotation for every premutation of units feels unnecessary.

This is true, but as it is, `AnnotatedQuantity` is unsafe because when used without specifying `Field(json_schema_extra={'unit': ...})` on a model, the validator will fail (and not at build time, but only when the first validation is performed!). Indeed, defining a new annotation for all units is probably cumbersome to do, but you still have to specify the unit in the `Field.json_schema_extra` argument anyway when using the annotation.

The alternative can be to do the following:

```python
from pint import Quantity

class Box(BaseModel):
    w: Annotated[Quantity, WithUnit('in')]
```

And have `WithUnit` implemented the similarly to the `ValidateThirdParty` from my previous comment.

I understand having access to `model_fields` gives more flexibility, but trying to implement such a feature is a bit more difficult that we can expect:

- Using the `Field()` function is not limited to Pydantic models. The same applies to Pydantic/stdlib dataclasses for instance. Dataclasses also have fields accessible under via the `__pydantic_fields__` property.
- While in your case, you might only care about using the third-party type in Pydantic models, this pattern will not work for [type adapters](https://docs.pydantic.dev/latest/concepts/type_adapter/). We may argue that this is also an issue for the exposed `field_name` (which is `None` when using type adapters), but `Field()` can also be used with them, meaning the following wouldn't work: `TypeAdapter(Annotated[AnnotatedQuantity, Field(json_schema_extra=...)])`.

So regarding the last point, if we were to expose something similar to `model_fields`, that would the the actual `FieldInfo` instance:

```python
def validate(value, info):
    unit = info.field_info.json_schema_extra['unit']
    # perform validation ...

class Model(BaseModel):
    f: <type> = Field(json_schema_extra={'unit': 'in'})
```

But note that this would still require work to be done to make it work with type adapters, as we don't really create `FieldInfo` instances internally for them.

**Comment by dennispg** [comment]
To clarify, I have revised the request here to expose `model_type` and not `model_fields` as I had previously.

I'm curious why say that the validator would fail, and if that is based on the assumption that the validator would fail without having unit parameter. In my case, I guard against the case when the units are not provided and validate in a different way if it isn't available. I imagine any validator already making use of `field_name` would do something like this as well. already

I do use the json schema in my case, and so using a WithUnit annotation would mean a duplication of units in both the annotation and in the Field. Not the end of the world I suppose, but still feels unnecessarily clunky.

In short, it is my perspective that any use case argument that can be made for `field_name` can also be made for providing `model_type`. The information is already there to be provided.

Thank you for taking the time to think through this with me! I hope you won't mind, but I am going to go ahead and submit my PR for review to further this discussion.

**Comment by Viicos** [comment]
> To clarify, I have revised the request here to expose `model_type` and not `model_fields` as I had previously.

Yes my bad, you can discard my first bullet point then.

> I'm curious why you say that the validator would fail, and if that is based on the assumption that the validator would fail without having the unit parameter.

If we take the example you gave in the OP:

```python
def validate_length(value: Any, info: ValidationInfo):
    #
    #   I want to see the json_schema_extra value from here
    #
    if info.field_name is not None and info.model_type is not None:
      field = info.model_type.model_fields[info.field_name]
        if field is not None and isinstance(field.json_schema_extra, dict):
          if field.json_schema_extra["absolute"]:
            return abs(value)
    return value

class SimpleThirdPartyType(BaseModel):
    length: int

AnnotatedModel = Annotated[SimpleThirdPartyType, BeforeValidator(validate_length)]

class SimpleOuterModel(BaseModel):
    value: AnnotatedModel = Field(json_schema_extra={"absolute": True})
```

`SimpleOuterModel` will build fine when you import it, but only when you validate data with it you will get a runtime exception if `json_schema_extra` is empty or missing the `absolute` value. Of course, you can safely inspect the `json_schema_extra` dict, and maybe using a default value for `absolute`, but still — conceptually — it feels weird to have validator build logic (i.e. determining _where_ to look for the `absolute` value) during validation.

> I do use the json schema in my case, and so using a WithUnit annotation would mean a duplication of units in both the annotation and in the Field.

The `WithUnit` takes care of adding the necessary JSON Schema info, so you don't need to duplicate it in the `Field()` function (see the `ValidateThirdParty.__get_pydantic_json_schema__` method — besides, I also gave an example where you can customize the JSON Schema even further based on the mode, `'validation'`/`'serialization'`). This approach is cleaner as you don't go back and forth between the `FieldInfo` instance and the validator function implementation at validation time.

> In short, it is my perspective that any use case argument that can be made for `field_name` can also be made for providing `model_type`. The information is already there to be provided.

A couple points where I believe this is not true:
- `field_name` is `str | None`. It is easier to account for both use cases, while in the case of `model_type`, this can either be a `BaseModel`, a dataclass, a typed dict. For each case, you'll have to look for the fields in a different attribute.
- Providing `model_type` will not work for type adapters, even though the following is supported by Pydantic: `TypeAdapter(Annotated[<typ>, Field(json_schema_extra(...)])` (see my second bullet point in my previous comment).

**Comment by dennispg** [comment]
Your points are well taken. For my use case and for the simplistic example provided here, you have convinced me that using a WithUnit sort of annotation would be a better and cleaner choice and I will actually be refactoring my own project that prompted this request now.

That said, it might just be that these examples aren't the best ones to make the case for providing access to `model_type`. As far how this affects the usage with TypeAdapter, I'm still not convinced that this is any different from `field_name`.

I do hear what you're saying about whether `model_type` a BaseModel, dataclass, or typed dict, but I suspect this may all be now wading into the philosophical discussion of why `field_name` was provided in the first place and what it's usefulness is. I'll leave it to you guys to decide what to do here, but I'll be happy to help or be included in any discussions, even if only as a fly on the wall.

Also, I will be more than happy to finish out the tests and CI issue fixes if you guys decide to move forward with this.

Thanks!

**Comment by Viicos** [comment]
> That said, it might just be that these examples aren't the best ones to make the case for providing access to `model_type`. As far how this affects the usage with TypeAdapter, I'm still not convinced that this is any different from `field_name`.

In fact, my point with type adapters does not really relate to the sole concept of providing `model_type`. If we only consider this addition, this is indeed no different from `field_name`. My argument was related to _how_ you can make use of `model_type`, and in your case showing that it wouldn't work when trying to fetch the `json_schema_extra`.

> why `field_name` was provided in the first place and what it's usefulness is.

I agree on the fact that it may be available or not shows that this isn't really ideal (and this is reflected by the fact that it is typed as `str | None`: unions in the typing world are generally a sign that the implementation may be flawed in some way). I don't know exactly why it was introduced in the first place, this was done here: https://github.com/pydantic/pydantic/pull/7542. Seems like [Samuel was pushing back on the feature in the first place](https://github.com/pydantic/pydantic/issues/7448#issuecomment-1721392042).

We'll soon discuss about a better API to support custom types (in particular, one that would avoid the annotated pattern so that you can just use the third party type as an annotation). Ideally, the API would allow accessing the `FieldInfo` instance and/or the annotated metadata, making it easier to support such use cases.

**Comment by dennispg** [comment]
From that discussion it looks like the purpose for it was simply to just be less restrictive. Until such time as that better API comes along, would it make sense then to merge my changes in service to that same purpose?

I am grateful to have been able to participate in this discussion so it won't hurt my feelings either way, but it sure would make me happy to be able contribute something back. 😁

**Comment by Viicos** [comment]
Thanks for being involved! This is the kind of discussion that we should see more often in open source. Designing APIs is hard, especially for libraries.

Without a real request for `model_type` (as if I understand correctly you are not going to rely on it), let's not add the feature for now. We can revisit later if this is requested again but the fact that `model_type` can be of different types is not really ideal.

By the way, if you are interested, this discussion looks like an example (maybe not the best) of attractive nuisances, you can read more in this [article](https://blog.ganssle.io/articles/2023/01/attractive-nuisances.html).

I'll close the issue and PRs, thanks for looking at the implementation still
